### ISSUE TITLE
regression: Generalized failures in tests related to postgres replication - Attempted to delay Capability

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#18686]
### What version of Materialize are you using?

ebb39fcd31

### How did you install Materialize?

Docker image

### What is the issue?

Multiple failures in postgres-related tests have started happening:

1. ~~A failure in the `cluster` tests~~
https://buildkite.com/materialize/tests/builds/53291#01876358-f080-4fda-9c26-d67f923d4ab0

2. A failure in all "Checks" tests that restart Mz or parts thereof:
https://buildkite.com/materialize/tests/builds/53276#01875e35-74c8-4b87-a8be-3be8c348596f
```
[2023-04-08T00:22:42Z] > INSERT INTO many_rows SELECT * FROM many_rows;
--
  | [2023-04-08T00:27:42Z] rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 76s
  | [2023-04-08T00:27:42Z] 2:1: error: preparing query failed: connection closed
  | [2023-04-08T00:27:42Z]      \|
  | [2023-04-08T00:27:42Z]    1 \|
  | [2023-04-08T00:27:42Z]    2 \| > INSERT INTO many_rows SELECT * FROM many_rows;
  | [2023-04-08T00:27:42Z]      \| ^
```
This is a panic:

```
platform-checks-materialized-1        | thread 'timely:work-1' panicked at 'Attempted to delay Capability { time: MzOffset { offset: 38682089 }, internal: "..." } to MzOffset { offset: 38680776 }, which is not beyond the capability's time.', /cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/134842a/timely/src/dataflow/operators/capability.rs:104:13
```

ci-regexp: Attempted to delay Capability

3.  A failure in `pg-cdc-resumption` that indicates the source postgres table may have been ingested multiple times(?):
https://buildkite.com/materialize/tests/builds/53276#01875e35-74ac-4ae3-9e54-1c6cdbc5b507

```
verify-data.td | 2m 0s
-- | --
  | [2023-04-08T00:17:41Z] > SELECT COUNT(*) FROM t1;
  | [2023-04-08T00:19:41Z] rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s
  | [2023-04-08T00:19:41Z] verify-data.td:10:1: error: non-matching rows: expected:
  | [2023-04-08T00:19:41Z] [["500000"]]
  | [2023-04-08T00:19:41Z] got:
  | [2023-04-08T00:19:41Z] [["1500000"]]
  | [2023-04-08T00:19:41Z] Poor diff:
  | [2023-04-08T00:19:41Z] + 1500000
  | [2023-04-08T00:19:41Z] - 500000
```

ci-regexp: verify-data.td:10:1: error: non-matching rows

**Comment by philip-stoev** [comment]
Examination of the CI runs and additional runs performed strongly indicate that at least No 2 is associated with this commit:

```
commit acc39ea1b9841fef17e42732754bc26e801a00ef
Author: Joseph Koshakow <koshy44@gmail.com>
Date:   Fri Apr 7 16:34:37 2023 -0400

    adapter: Refactor cascading delete statements (#18651)

    Previously, the adapter would collect all the IDs of objects to delete
    as part of a cascading delete during sequencing. This commit refactors
    this logic so the IDs are collected during planning. This allows us to
    include those IDs as part of RBAC checks.

    Additionally, we had to move the validation that no user was directly
    modifying a linked cluster to the planner.

    Fixes #18551
```

It has not happened before but is quite reproducible with this commit. The problem is that it does not reproduce well on a single machine, one needs to run the entire CI to get an instance. To get a CI run against a specific commit, you may need to set the `BUILDKITE_TAG=...` env variable in the CI, otherwise the `mkpipeline.py` script may decide to prune the entire buildkite pipeline.

Item No 1 has been since once in the commit just prior, so may be a separate thing, I will bisect it separartely.

**Comment by philip-stoev** [comment]
@jkosh44 FYI

**Comment by sploiselle** [comment]
Just wanted to note that the failure in cluster tests (https://buildkite.com/materialize/tests/builds/53291#01876358-f080-4fda-9c26-d67f923d4ab0) you linked as issue 1. was known to be caused by #18644, and was almost certainly fixed by #18672

**Comment by jkosh44** [comment]
For issue (2), I also tried to extract all the SQL queries run. The queries that start with `>` are run in Materialize directly. The queries that start with `>>` are either run in an upstream PostgreSQL or in Materialize on the internal port, There's a preceding line of the format `Connecting to PostgreSQL server at ...` that helps differentiate the two possibilities. 
[test_sql.txt](https://github.com/MaterializeInc/materialize/files/11191031/test_sql.txt)

Interestingly, there isn't a single `DROP ... CASCADE` run. Additionally there's no `CREATE OR REPLACE` run. Below are all of the `DROP` statements extracted from the SQL:
```
> DROP TABLE drop_table1;
> DROP INDEX drop_index_index1;
> DROP TABLE drop_table2;
> DROP INDEX drop_index_table_primary_idx;
> DROP INDEX drop_index_index2;
```

Here are all the statements involving those objects in order:
```
> CREATE TABLE drop_table1 (f1 INTEGER);
> INSERT INTO drop_table1 VALUES (1);
> CREATE TABLE drop_table2 (f1 INTEGER);
> INSERT INTO drop_table2 VALUES (1);
> CREATE INDEX drop_index_index1 ON drop_index_table (f1, f2);
> DROP TABLE drop_table1;
> DROP INDEX drop_index_index1;
> CREATE INDEX drop_index_index2 ON drop_index_table (f1, f2);
>> ALTER TABLE drop_table2 OWNER TO materialize;
> DROP TABLE drop_table2;
>> ALTER INDEX drop_index_table_primary_idx OWNER TO materialize;
>> ALTER INDEX drop_index_index2 OWNER TO materialize;
> DROP INDEX drop_index_table_primary_idx;
> DROP INDEX drop_index_index2;
```

It's important to note that none of the tables have indexes on them, which would have implicitly turned them into cascading drops.

It is likely that before the test was run there was a `DROP DATABASE materialize CASCADE` run to clean up the environment.

**Comment by jkosh44** [comment]
For issue (2), so far I've noticed the same issue in the following tests:
- `checks-kill-clusterd-storage`
- `checks-restart-entire-mz`
- `checks-restart-environmentd-clusterd-storage`

(This build has the most examples that I've seen: https://buildkite.com/materialize/tests/builds/53276).

It always happens in a test that restarts some component of Materialize, which suggests to me that we're not properly writing something down on disk. The fact that it's reproducible in `checks-kill-clusterd-storage` further suggest to me that a `clusterd` used by storage is specifically not properly writing something down.

The timestamp type used in the panic is `mz_storage_client::types::sources::MzOffset` and the stack trace originates here:
```
  21: <mz_storage::storage_state::Worker<timely_communication::allocator::generic::Generic>>::run_client
  22: <mz_storage::storage_state::Worker<timely_communication::allocator::generic::Generic>>::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0639d3bfc2b2748a1-1/materialize/tests/src/storage/src/storage_state.rs:411:21
  23: <mz_storage::server::Config as mz_cluster::types::AsRunnableWorker<mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse>>::build_and_run::<timely_communication::allocator::generic::Generic>
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0639d3bfc2b2748a1-1/materialize/tests/src/storage/src/server.rs:105:9
  24: <mz_cluster::server::ClusterClient<mz_service::client::Partitioned<mz_service::local::LocalClient<mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse, std::thread::Thread>, mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse>, mz_storage::server::Config, mz_storage_client::client::StorageCommand, mz_storage_client::client::StorageResponse>>::build_timely::{closure#0}::{closure#1}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0639d3bfc2b2748a1-1/materialize/tests/src/cluster/src/server.rs:179:13
```
which further points to a storage `clusterd` process.

**Comment by jkosh44** [comment]
@sploiselle @benesch Based on the stack trace for issue (2), would it be accurate to say that the panic is coming from either a source or a materialized view?

**Comment by jkosh44** [comment]
For issue (2) I just reproduced it in the following commit:

```
commit 5ca9b7be8923788e4d72a39a608298216bb10732
Author: Sean Loiselle <sean@materialize.io>
Date:   Fri Apr 7 11:35:23 2023 -0400

    pg sources: add timeout to reading replication stream
```

https://buildkite.com/materialize/tests/builds/53338#01876bdc-8515-4675-90f5-fd0010f981b0

Which is an earlier commit than the one listed earlier:
```
commit acc39ea1b9841fef17e42732754bc26e801a00ef
Author: Joseph Koshakow <koshy44@gmail.com>
Date:   Fri Apr 7 16:34:37 2023 -0400

    adapter: Refactor cascading delete statements (#18651)

    Previously, the adapter would collect all the IDs of objects to delete
    as part of a cascading delete during sequencing. This commit refactors
    this logic so the IDs are collected during planning. This allows us to
    include those IDs as part of RBAC checks.

    Additionally, we had to move the validation that no user was directly
    modifying a linked cluster to the planner.

    Fixes #18551
```

I can't say for certain if it was introduced by `5ca9b7be8923788e4d72a39a608298216bb10732`, but it definitely wasn't introduced by `acc39ea1b9841fef17e42732754bc26e801a00ef`. I'll have to keep trying to repro with older commits to try and find the earliest commit that can repro.

**Comment by guswynn** [comment]
This reminds me of: https://github.com/MaterializeInc/materialize/issues/18329. There might be some issue in the pg source that causes re-ingestion after restarts, though in the past it didnt cause a capability panic. There is also prior art on capability panics being related to dataloss or duplication

**Comment by sploiselle** [comment]
Fixed by #18697
